src/CoreRCON/RCON.cs

```diff
@@ -175,52 +197,42 @@ await writer.CompleteAsync()
         /// </summary>
         /// <param name="reader"></param>
         /// <returns>Consumer Task</returns>
-        async Task ReadPipeAsync(PipeReader reader)
+        async Task ReadPipeAsync(PipeReader reader, CancellationToken cancellationToken)
         {
             try
             {
                 while (true)
                 {
-                    ReadResult result = await reader.ReadAsync()
+                    ReadResult result = await reader.ReadAsync(cancellationToken)
                         .ConfigureAwait(false);
                     ReadOnlySequence<byte> buffer = result.Buffer;
                     SequencePosition packetStart = buffer.Start;
 
+                    if (cancellationToken.IsCancellationRequested)
+                        break;
+
                     if (buffer.Length < 4)
                     {
                         if (result.IsCompleted)
                         {
                             break;
                         }
+
                         reader.AdvanceTo(packetStart, buffer.End);
                         continue;
                         // Complete header not yet received
                     }
+
                     int size = BitConverter.ToInt32(buffer.Slice(packetStart, 4).ToArray(), 0);
                     if (buffer.Length >= size + 4)
                     {
                         // Get packet end positions 
                         SequencePosition packetEnd = buffer.GetPosition(size + 4, packetStart);
                         byte[] byteArr = buffer.Slice(packetStart, packetEnd).ToArray();
                         RCONPacket packet = RCONPacket.FromBytes(byteArr);
-
-                        if (packet.Type == PacketType.AuthResponse)
```

Comment:
I really like this change moving authenticate to its own Task awaited on connect 👍